src/server.rs

```diff
@@ -222,6 +221,8 @@ impl Server {
                 }
                 self.data.clear();
             }
+        } else if socket.may_send() {
```

Comment:
I think this would make more sense from a reader's perspective if we wrote it as a separate conditional. Pull it down a line and change it to 
```rust
if socket.may_send() && !socket.may_recv() {
    socket.close();
}
```

That makes it more explicit why we're closing the socket.